The aim of this exercise is to train you in debugging networks using the good old print function and also tensorboard. To simulate poor training, we will train a multilayer perceptron using the CIFAR data.

1. Use the CIFAR data set reader from the first homework and read the CIFAR-10 files again. 
2. Apply random noise to the image 
3. Convert the image to float and scale to [0.0, 1.0] by dividing the pixel values by the highest pixel value.
4. Convert all labels to onehot encoding
5. Build a 3-layer multilayer perceptron of size [512, 256, 128]. 
6. Create a tensorboard summary for plotting the histogram of the weights of the three layers.
7. Also write the cost / loss at the end of each epoch to tensorboard.
8. Train the network with learning rates of [0.1, 0.01, 0.001]. You will notice that the network will not converge well.
9. Submit the snapshot of the histograms for the three learning rates. Describe your observations.



In [1]:
import pickle
import glob
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf 
import pandas as pd 
import random
import datetime

random.seed(3457)

#1. Use the CIFAR data set reader from the first homework and read the CIFAR-10 files again. 
files = glob.glob(
    "/Users/omerorhan/Desktop/UCSC/Deep Learning/cifar-10-batches-py/data_batch*")

def unpickle(file, types):
    with open(file, 'rb') as fo:
        return pickle.load(fo, encoding='bytes').get(types)

data = np.concatenate([unpickle(x, b'data') for x in files])
labels = np.concatenate([unpickle(x, b'labels') for x in files])




/anaconda3/envs/tf-cpu/lib/python3.5/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [3]:
#2. Apply random noise to the image 
mean = 1   # some constant
std = 1    # some constant (standard deviation)
noisy_data = data + np.random.normal(mean, std, data.shape)

#3. Convert the image to float and scale to [0.0, 1.0] by dividing the pixel values by the highest pixel value.
noisy_data = noisy_data.astype(float)
noisy_data = np.divide(noisy_data, np.amax(noisy_data[0]))

noisy_data = np.float32(noisy_data)


In [4]:
#4. Convert all labels to onehot encoding
n_values = np.max(labels) + 1
print(type(labels))
labelsencoded = np.eye(n_values)[labels]
labelsencoded = np.float32(labelsencoded)

print(labelsencoded[25])

<class 'numpy.ndarray'>
[0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]


In [5]:
# Hyper-parameters

learningrate = 0.001
nepochs = 100
batch_size = 128
#128,3072
nhidden1 = 512
nhidden2 = 256 
nhidden3 = 128 
ninput = noisy_data[0].flatten().shape[0]
noutput = 10

X = tf.placeholder(tf.float32, [batch_size,ninput])
Y = tf.placeholder(tf.float32, [batch_size, noutput])


print(X)
print(Y)

Tensor("Placeholder:0", shape=(128, 3072), dtype=float32)
Tensor("Placeholder_1:0", shape=(128, 10), dtype=float32)


In [6]:
#5. Build a 3-layer multilayer perceptron of size [512, 256, 128]. 

weights = \
{
        'h1': tf.Variable(tf.random_normal([ninput, nhidden1])),
        'h2': tf.Variable(tf.random_normal([nhidden1, nhidden2])),
        'h3': tf.Variable(tf.random_normal([nhidden2, nhidden3])),
        'out': tf.Variable(tf.random_normal([nhidden3, noutput]))
}

biases = \
{
    'b1': tf.Variable(tf.random_normal([nhidden1])),
    'b2': tf.Variable(tf.random_normal([nhidden2])),
    'b3': tf.Variable(tf.random_normal([nhidden3])),
    'out': tf.Variable(tf.random_normal([noutput]))
}

def multiperceptron(x):
    l1 = tf.nn.sigmoid(tf.add(tf.matmul(x, weights['h1']), biases['b1']))
    l2 = tf.nn.sigmoid(tf.add(tf.matmul(l1, weights['h2']), biases['b2']))
    l3 = tf.nn.sigmoid(tf.add(tf.matmul(l2, weights['h3']), biases['b3']))
    outl = tf.nn.sigmoid(tf.add(tf.matmul(l3, weights['out']), biases['out']))
    return outl

model = multiperceptron(X)


Instructions for updating:
Colocations handled automatically by placer.


In [7]:

loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits = model, labels=Y))
optimizer = tf.train.AdamOptimizer(learning_rate=learningrate)
train_min = optimizer.minimize(loss)

correct_prediction = tf.equal(tf.argmax(model, 1), tf.argmax(Y, 1))
# argmax tells what index is the max in the y hat
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
tf.summary.scalar("accuracy", accuracy)

#6. Create a tensorboard summary for plotting the histogram of the weights of the three layers.
# Add to tensorboard summary
tf.summary.histogram("weight_1",weights['h1'])
tf.summary.histogram("weight_2",weights['h2'])
tf.summary.histogram("weight_3",weights['h3'])
tf.summary.scalar("loss", loss)

merged_summary_op = tf.summary.merge_all()
filename="/Users/omerorhan/Desktop/UCSC/Deep Learning/DeepLearningUCSC/venv/homework/Assignments" \
         "/logs/run"+datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S_learningrate"+str(learningrate))
writer = tf.summary.FileWriter(filename, tf.get_default_graph())
init = tf.global_variables_initializer()


Instructions for updating:

Future major versions of TensorFlow will allow gradients to flow
into the labels input on backprop by default.

See `tf.nn.softmax_cross_entropy_with_logits_v2`.



In [8]:
import random
def getbatch(alldata, alllabels, batch_size = 16):
    nlabels = len(alllabels)
    number_of_batches = nlabels//batch_size # TODO: Change 100 to nlabels
    for batch_number in range(number_of_batches):
        rand_index = [random.randrange(0, nlabels) for i in range(batch_size)]
        batch_x = alldata[rand_index]
        batch_y = np.array([alllabels[idx] for idx in rand_index])
        yield (batch_x, batch_y)
 


In [9]:

with tf.Session() as sess:
    sess.run(init)
    
    for epoch in range(nepochs):
        for batch_x, batch_y in getbatch(noisy_data, labelsencoded, batch_size):
            sess.run(train_min, feed_dict={X:batch_x, Y:batch_y})
        losscalc, accuracycalc, merged_summary = \
                      sess.run([loss, accuracy, merged_summary_op], feed_dict={X:batch_x, Y:batch_y})
        writer.add_summary(merged_summary, epoch) 
        
        print("Epoch: %d, Loss: %0.4f, Accuracy: %0.4f"%(epoch, losscalc, accuracycalc))
writer.close()

Epoch: 0, Loss: 2.1841, Accuracy: 0.0938


Epoch: 1, Loss: 2.0179, Accuracy: 0.3203


Epoch: 2, Loss: 2.0523, Accuracy: 0.2891


Epoch: 3, Loss: 2.0568, Accuracy: 0.3281


Epoch: 4, Loss: 2.0231, Accuracy: 0.3906


Epoch: 5, Loss: 1.9734, Accuracy: 0.4219


Epoch: 6, Loss: 2.0112, Accuracy: 0.3516


Epoch: 7, Loss: 2.0132, Accuracy: 0.3828


Epoch: 8, Loss: 1.9717, Accuracy: 0.3594


Epoch: 9, Loss: 1.9849, Accuracy: 0.3594


Epoch: 10, Loss: 1.9560, Accuracy: 0.3750


Epoch: 11, Loss: 1.9740, Accuracy: 0.4297


Epoch: 12, Loss: 1.9534, Accuracy: 0.4297


Epoch: 13, Loss: 1.9460, Accuracy: 0.4141


Epoch: 14, Loss: 1.9510, Accuracy: 0.4141


Epoch: 15, Loss: 1.9617, Accuracy: 0.4141


Epoch: 16, Loss: 1.8918, Accuracy: 0.4297


Epoch: 17, Loss: 1.9079, Accuracy: 0.4141


Epoch: 18, Loss: 1.9372, Accuracy: 0.3828


Epoch: 19, Loss: 1.9290, Accuracy: 0.3594


Epoch: 20, Loss: 1.9519, Accuracy: 0.3359


Epoch: 21, Loss: 1.9106, Accuracy: 0.3984


Epoch: 22, Loss: 1.9724, Accuracy: 0.3672


Epoch: 23, Loss: 1.9516, Accuracy: 0.3594


Epoch: 24, Loss: 1.9494, Accuracy: 0.3516


Epoch: 25, Loss: 1.9411, Accuracy: 0.3594


Epoch: 26, Loss: 1.9135, Accuracy: 0.3906


Epoch: 27, Loss: 1.9334, Accuracy: 0.3516


Epoch: 28, Loss: 1.9111, Accuracy: 0.4453


Epoch: 29, Loss: 1.9916, Accuracy: 0.3203


Epoch: 30, Loss: 1.9320, Accuracy: 0.4141


Epoch: 31, Loss: 1.8811, Accuracy: 0.3672


Epoch: 32, Loss: 1.9175, Accuracy: 0.3828


Epoch: 33, Loss: 1.9431, Accuracy: 0.3750


Epoch: 34, Loss: 1.9858, Accuracy: 0.3672


Epoch: 35, Loss: 1.9538, Accuracy: 0.4141


Epoch: 36, Loss: 1.8495, Accuracy: 0.5469


Epoch: 37, Loss: 1.9013, Accuracy: 0.4531


Epoch: 38, Loss: 1.9437, Accuracy: 0.4375


Epoch: 39, Loss: 1.9136, Accuracy: 0.4297


Epoch: 40, Loss: 1.9462, Accuracy: 0.3125


Epoch: 41, Loss: 1.8529, Accuracy: 0.5078


Epoch: 42, Loss: 1.8793, Accuracy: 0.4375


Epoch: 43, Loss: 1.9474, Accuracy: 0.3750


Epoch: 44, Loss: 1.9311, Accuracy: 0.3438


Epoch: 45, Loss: 1.8663, Accuracy: 0.4609


Epoch: 46, Loss: 1.9444, Accuracy: 0.4062


Epoch: 47, Loss: 1.9354, Accuracy: 0.4453


Epoch: 48, Loss: 1.8613, Accuracy: 0.4531


Epoch: 49, Loss: 1.9519, Accuracy: 0.3984


Epoch: 50, Loss: 1.9288, Accuracy: 0.3516


Epoch: 51, Loss: 1.7872, Accuracy: 0.5312


Epoch: 52, Loss: 1.9143, Accuracy: 0.4531


Epoch: 53, Loss: 1.8766, Accuracy: 0.4609


Epoch: 54, Loss: 1.8677, Accuracy: 0.4375


Epoch: 55, Loss: 1.8584, Accuracy: 0.3984


Epoch: 56, Loss: 1.8758, Accuracy: 0.4297


Epoch: 57, Loss: 1.8747, Accuracy: 0.4375


Epoch: 58, Loss: 1.8956, Accuracy: 0.4062


Epoch: 59, Loss: 1.9183, Accuracy: 0.3828


Epoch: 60, Loss: 1.9663, Accuracy: 0.4375


Epoch: 61, Loss: 1.9005, Accuracy: 0.3594


Epoch: 62, Loss: 1.9116, Accuracy: 0.4219


Epoch: 63, Loss: 1.9124, Accuracy: 0.3594


Epoch: 64, Loss: 1.9142, Accuracy: 0.4062


Epoch: 65, Loss: 1.8689, Accuracy: 0.3906


Epoch: 66, Loss: 1.8682, Accuracy: 0.4375


Epoch: 67, Loss: 1.8657, Accuracy: 0.4844


Epoch: 68, Loss: 1.8658, Accuracy: 0.4609


Epoch: 69, Loss: 1.9281, Accuracy: 0.3750


Epoch: 70, Loss: 1.8910, Accuracy: 0.3906


Epoch: 71, Loss: 1.8609, Accuracy: 0.4375


Epoch: 72, Loss: 1.8764, Accuracy: 0.4688


Epoch: 73, Loss: 1.8657, Accuracy: 0.4453


Epoch: 74, Loss: 1.8601, Accuracy: 0.4688


Epoch: 75, Loss: 1.8525, Accuracy: 0.4375


Epoch: 76, Loss: 1.8674, Accuracy: 0.4141


Epoch: 77, Loss: 1.8580, Accuracy: 0.4531


Epoch: 78, Loss: 1.9162, Accuracy: 0.4219


Epoch: 79, Loss: 1.8841, Accuracy: 0.4297


Epoch: 80, Loss: 1.9311, Accuracy: 0.4219


Epoch: 81, Loss: 1.8577, Accuracy: 0.5156


Epoch: 82, Loss: 1.8453, Accuracy: 0.4375


Epoch: 83, Loss: 1.9367, Accuracy: 0.3984


Epoch: 84, Loss: 1.8597, Accuracy: 0.4688


Epoch: 85, Loss: 1.8777, Accuracy: 0.4688


Epoch: 86, Loss: 1.8981, Accuracy: 0.4062


Epoch: 87, Loss: 1.8629, Accuracy: 0.5078


Epoch: 88, Loss: 1.8586, Accuracy: 0.4688


Epoch: 89, Loss: 1.7906, Accuracy: 0.4531


Epoch: 90, Loss: 1.8478, Accuracy: 0.4922


Epoch: 91, Loss: 1.9074, Accuracy: 0.4531


Epoch: 92, Loss: 1.8294, Accuracy: 0.4766


Epoch: 93, Loss: 1.8541, Accuracy: 0.3516


Epoch: 94, Loss: 1.8681, Accuracy: 0.4062


Epoch: 95, Loss: 1.8742, Accuracy: 0.4141


Epoch: 96, Loss: 1.8659, Accuracy: 0.4453


Epoch: 97, Loss: 1.9262, Accuracy: 0.4062


Epoch: 98, Loss: 1.8161, Accuracy: 0.5156


Epoch: 99, Loss: 1.8920, Accuracy: 0.4688


### Accuracy(learning rate 0.1):
In the figure below, you will notice that the accuracy is around 10%. 
Due to learning rate(0.1), accuracy is very bad.

<img src="hm2_img/lr0.1accuracy.png" height="200px">


### Loss(learning rate 0.1):
In the figure below, you will notice that the loss is around 2.2 - 2.38%. 

<img src="hm2_img/lr0.1loss.png" height="200px">

### Accuracy(learning rate 0.01):
In the figure below, you will notice that the accuracy is around 10%. 
Due to learning rate(0.01), accuracy is very bad.

<img src="hm2_img/lr0.01accuracy.png" height="200px">


### Loss(learning rate 0.01):
In the figure below, you will notice that the loss is around 2.2 - 2.38%. 

<img src="hm2_img/lr0.01loss.png" height="200px">


### Accuracy(learning rate 0.001):
In the figure below, you will notice that the accuracy is around 10%. 
Due to learning rate(0.001), accuracy is very bad.

<img src="hm2_img/lr0.001accuracy.png" height="200px">


### Loss(learning rate 0.001):
In the figure below, you will notice that the loss is around 2.2 - 2.38%. 

<img src="hm2_img/lr0.001loss.png" height="200px">
